Add refactor migration guide (#495)
Green-Software-Foundation/if: Refactor-migration-guide.md

# Managing the IF refactor

We recently completed a major refactor of the entire IF codebase, moving from an object oriented to functional programmign style.

If you are a IF user or developer, there are some changes you must be aware of resulting from a recent refactor of the IF codebase. This guide will help you to update your processes to integrate your work with the newly refactored IF.

## Running IF

There have been some name changes to the CLI, specifically:

> - `impact-engine` --> `if`

Reviewer: Nit: Since`if` was changed to `ie` this should be adjusted in the guide as well.

Author: ```suggestion
- `impact-engine` --> `ie`
```

>     The command line tool has been renamed from `impact-engine` to simply `if`. This means that to invoke the Impact Framework on the command line you simply use 

Author: ```suggestion
    The command line tool has been renamed from `impact-engine` to simply `ie`. This means that to invoke the Impact Framework on the command line you simply use 
```

    ```

>     if ...

Author: ```suggestion
    ie ...
```

    ```
-  `impl` --> `manifest` 
    We have deprecated the original `impl` and `ompl` terminology across all our repositories and on the command line. Now, to pass a manifest file to IF, you use the `--manifest` command, as follows:

    ```sh

>     if --manifest <path-to-manifest>

Author: ```suggestion
    ie --manifest <path-to-manifest>
```

    ```

- `ompl` --> `output`
 

>     We have deprecated the original `impl` and `ompl` terminology across all our repositories and on the command line. Now, todefine a savepath for your output file, you use the `--output` command, as follows:

Author: ```suggestion
    We have deprecated the original `impl` and `ompl` terminology across all our repositories and on the command line. Now, to define a savepath for your output file, you use the `--output` command, as follows:
```

    ```sh

>     if --manifest <path-to-manifest> --output <savepath>

Author: ```suggestion
    ie --manifest <path-to-manifest> --output <savepath>
```

    ```

## Outputs

We currently default to exporting output data to the console only. If you want to export to a file, you have to add a small piece of config inside your manifest, as follows:

```yaml
initalize:
  plugins: ...
  outputs: ['yaml']
```

You then provide your savepatrh (without file extension) to the `--output` command on the CLI.

The available options are currently `csv`, `yaml` or `log`.

### Summary of how to run the refactored IF 

As before, you can install IF from our npm package using

```
npm i @grnsft/if
```

Then run IF using the following command:

```sh

> if --manifest <path-to-manifest>

Author: ```suggestion
ie --manifest <path-to-manifest>
```

```

This will dump the output to the console. If you want to save the output to a yaml file, provide a savepath to the `--output` command:

```sh

> if --manifest <path-to-manifest> --output <savepath>

Author: ```suggestion
ie --manifest <path-to-manifest> --output <savepath>
```

```

If you want to clone the source code and install and run a local copy of IF, you can do so using:

```sh
git clone https://github.com/Green-Software-Foundation/if &&
cd if &&
npm i
```

Then run IF using the following command:

```sh
npm run if -- --manifest <path-to-manifest>
```

## Manifest files

There have also been some changes to the structure of manifest files. Some of these are simple renaming changes, others are more functional.

- **Rename `graph` -> `tree`**
    The `graph` section of the manifest file is now renamed to `tree`. This is just to help us stay consistent in our metaphors and provide a more familiar naming convention for the data beneath.

- **Use plugin name as key in `Initialize` block**
    In the previous version of IF, the plugins were organized into an array each having a `name` key, with the plugin name as the value. In the refactored IF, we use the name as the key identifying the plugin. For example, this is the OLD way:
    
    ```yaml
    initialize:
      plugins:
        - name: ccf

>           method: CloudCarbonFootprint

Author: ```suggestion
          model: CloudCarbonFootprint
```

>           method: CloudCarbonMetadataModel

Reviewer: ```suggestion
          model: CloudCarbonMetadataModel
```

Author: nope, it should be `method` now

Reviewer: Yes, but that is the old one, isn't it?

Author: Oh, it's the value you;re referring to, not the key. Yes, it should be
```suggestion
          method: CloudCarbonFootprint
```

Reviewer: No, sorry for the confusion: I corrected the example for the "OLD way" (see the context of the example). And previously you used the `model` key.
